File: Leetcode/206. Reverse Linked List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
    if(!head) return head;
    let myStack: ListNode[] = [];
    let cur: ListNode = head;
    
    while(cur) {
        myStack.push(cur);
        cur = cur.next;
    }
    
    let newHead: ListNode = myStack[myStack.length - 1];
    cur = newHead;
    myStack.pop();
        
    for(let i: number = myStack.length - 1; i>=0; i--) {
        cur.next = myStack[i];
        myStack.pop();
        cur = cur.next;
    }
    
    cur.next = null;
    
    return newHead;
};

File: Leetcode/206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head) return head;
        
        stack<ListNode*> s;
        ListNode *cur = head;
        
        while(cur) {
            s.push(cur);
            cur = cur->next;
        }
        
        ListNode *newHead = s.top();
        s.pop();
        cur = newHead;
        
        while(!s.empty()) {
            cur->next = s.top();
            s.pop();
            cur = cur->next;
        }
        cur->next = nullptr;
        
        return newHead;
    }
};
